reascripts/ReaSpeech/source/CurlRequest.lua

```diff
@@ -0,0 +1,368 @@
+--[[
+
+  CurlRequest.lua - wrapper interface for curl
+
+]]--
+
+CurlRequest = Polo {
+  DEFAULT_HTTP_METHOD = 'GET',
+  DEFAULT_HEADERS = {
+    ['accept'] = 'application/json',
+  },
+  DEFAULT_CURL_TIMEOUT = 0,
+  BASE_CURL_OPTIONS = {
+    '--http1.1', -- force HTTP/1.1
+    '-i',        -- include headers in output
+  },
+  DEFAULT_EXTRA_CURL_OPTIONS = {},
+  DEFAULT_ERROR_HANDLER = function(_msg) end,
+  DEFAULT_TIMEOUT_HANDLER = function() end,
+}
+
+function CurlRequest.async(options)
+  options.use_async = true
+  options.output_file = Tempfile:name()
+  options.sentinel_file = Tempfile:name()
+  options.progress_file = Tempfile:name()
+  options.extra_curl_options = options.extra_curl_options or {}
+
+  local extra_extra_curl_options = {
+    '-o ' .. options.output_file,
+    '--write-out "%output{' .. options.sentinel_file .. '}done"',
+    '--stderr ' .. options.progress_file,
+    '-#',
+
+    -- useful for testing transfer progress
+    -- '--limit-rate 2M',
+  }
+
+  for _, o in ipairs(extra_extra_curl_options) do
+    table.insert(options.extra_curl_options, o)
+  end
+
+  return CurlRequest.new(options)
+end
+
+function CurlRequest:init()
+  assert(self.url, 'missing url')
+
+  self.query_data = self.query_data or {}
+  self.http_method = self.http_method or self.DEFAULT_HTTP_METHOD
+  self.headers = self.headers or self.DEFAULT_HEADERS
+  self.curl_timeout = self.curl_timeout or self.DEFAULT_CURL_TIMEOUT
+  self.extra_curl_options = self.extra_curl_options or self.DEFAULT_EXTRA_CURL_OPTIONS
+  self.error_handler = self.error_handler or self.DEFAULT_ERROR_HANDLER
+  self.timeout_handler = self.timeout_handler or self.DEFAULT_TIMEOUT_HANDLER
+end
+
+function CurlRequest:execute()
+  local command = self:build_curl_command()
+
+  app:debug('CurlRequest: ' .. command)
```

Comment:
For a later date: it would be nice to avoid the dependency on `app`. I could see calling something like this in the `init` method:

```
Logging.init(self, 'CurlRequest')
```

which would add `self:log()` and `self:debug()` methods to the current object, automatically prefixing messages with the class name.

Reply:
i support this

Reply:
added issue https://github.com/TeamAudio/reaspeech/issues/79